From be11bc1ac9bc2b549dafaeace0ae3a80f5c9bcbb Mon Sep 17 00:00:00 2001
From: rmunjewar <rmunjewar@maxlinear.com>
Date: Thu, 11 Jul 2024 11:17:41 +0530
Subject: [PATCH] WLANRTSYS-82700 : iwlwav-hostap-ng : [AFC][ACS] BW
 upgrade/downgrade porting

Porting the changes to backup the user configured CW and restore it before candidiate creation

Porting changes from  commit
9518d75fd1a WLANRTSYS-80227 : iwlwav-hostap : BW downgrade/upgrade support in ACS WAV700 in 6GHz

FIXUP for 	WLANRTSYS-50170	Smart ACS implementation

Signed-off-by: RUSHABH MUNJEWAR <rmunjewar@maxlinear.com>
---
 hostapd/ctrl_iface.c     |  4 ++++
 hostapd/mxl_ctrl_iface.c |  3 +++
 src/ap/hostapd.c         |  7 ++++++-
 src/ap/mxl_hostapd.c     | 11 +++++++++++
 src/ap/mxl_hostapd.h     |  4 ++++
 5 files changed, 28 insertions(+), 1 deletion(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 5f289fb58..b001aca06 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -1666,6 +1666,10 @@ static int hostapd_ctrl_iface_update_reload(struct hostapd_data *hapd, char *cmd
 	hostapd_ctrl_iface_disable(iface);
 
 	hostapd_ctrl_iface_config_update(conf, cmd);
+#ifdef CONFIG_VENDOR_MXL
+	if (os_strstr(cmd, "bandwidth="))
+		mxl_hostapd_backup_user_conf_cw(iface);
+#endif /* CONFIG_VENDOR_MXL */
 
 	for (j = 0; j < iface->num_bss; j++) {
 		iface->bss[j]->conf->start_disabled = 0;
diff --git a/hostapd/mxl_ctrl_iface.c b/hostapd/mxl_ctrl_iface.c
index 683392021..fd28a758b 100644
--- a/hostapd/mxl_ctrl_iface.c
+++ b/hostapd/mxl_ctrl_iface.c
@@ -3156,6 +3156,9 @@ int mxl_hostapd_ctrl_iface_set(struct hostapd_data *hapd, char *cmd, int *ret)
 		else {
 			*ret = -1;
 		}
+	} else if ((os_strcmp(cmd, "vht_oper_chwidth") == 0) || (os_strcmp(cmd, "he_oper_chwidth") == 0)
+			   || (os_strcmp(cmd, "eht_oper_chwidth") == 0)) {
+		mxl_hostapd_backup_user_conf_cw(hapd->iface);
 	/* add new config item here as "} else if...{" */
 	} else
 		return 0;
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 07493230f..294fdb424 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -3950,6 +3950,9 @@ struct hostapd_iface * hostapd_init(struct hapd_interfaces *interfaces,
 		hapd->msg_ctx = hapd;
 	}
 
+#ifdef CONFIG_VENDOR_MXL
+	mxl_hostapd_init(hapd_iface);
+#endif /* CONFIG_VENDOR_MXL */
 	return hapd_iface;
 
 fail:
@@ -4879,7 +4882,9 @@ int hostapd_add_iface(struct hapd_interfaces *interfaces, char *buf)
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
 	if (start_ctrl_iface(hapd_iface) < 0)
 		goto fail;
-
+#ifdef CONFIG_VENDOR_MXL
+	mxl_hostapd_backup_user_conf_cw(hapd_iface);
+#endif /* CONFIG_VENDOR_MXL */
 	wpa_printf(MSG_INFO, "Add interface '%s'",
 		   hapd_iface->conf->bss[0]->iface);
 
diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index 62411b7fb..cb6a51b59 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -2956,3 +2956,14 @@ void mxl_hostapd_set_conf_from_freq_params(struct hostapd_iface *iface, const st
 	hostapd_set_oper_centr_freq_seg1_idx(iface->conf, seg1_idx);
 	hostapd_set_oper_chwidth(iface->conf, bw);
 }
+
+void mxl_hostapd_backup_user_conf_cw(struct hostapd_iface *iface)
+{
+	iface->mxl_iface.acs_user_config_cw = hostapd_get_oper_chwidth(iface->conf);
+	iface->mxl_iface.acs_user_config_op_class = iface->conf->op_class;
+}
+
+void mxl_hostapd_init(struct hostapd_iface * hapd_iface)
+{
+	mxl_hostapd_backup_user_conf_cw(hapd_iface);
+}
diff --git a/src/ap/mxl_hostapd.h b/src/ap/mxl_hostapd.h
index 094fce725..8e3adcc14 100644
--- a/src/ap/mxl_hostapd.h
+++ b/src/ap/mxl_hostapd.h
@@ -70,6 +70,8 @@ struct hostapd_failsafe_freq_params {
 struct mxl_hostapd_iface {
 	u32 sta_aid_link[MXL_AID_WORDS];
   s8 stas_free_reason; /* the reason for all stations remove */
+	int acs_user_config_cw;
+	int acs_user_config_op_class;
 	int allow_scan_during_cac;
 	struct hostapd_failsafe_freq_params failsafe;
 	u8 rbm; /* Radar Bit Map */
@@ -209,6 +211,8 @@ int mxl_hostapd_interface_update_power_regd(struct hostapd_data *hapd, int he_6g
 const char * mxl_he_6ghz_pwr_mode_text(enum he_6ghz_ap_type pwr_mode);
 int mxl_hostapd_no_ir_channel_list_updated(struct hostapd_iface *iface);
 int mxl_setup_interface2(struct hostapd_iface *iface);
+void mxl_hostapd_backup_user_conf_cw(struct hostapd_iface *iface);
+void mxl_hostapd_init(struct hostapd_iface * hapd_iface);
 
 void mxl_mlo_hostapd_interface_deinit(struct hostapd_iface *iface);
 void mxl_hostapd_remove_mld_vap(struct hostapd_iface *iface, unsigned int idx);
-- 
2.43.0

